### PR TITLE
feat: add a find in files panel on the GUI

### PR DESCRIPTION
**Pull request by smnppKDAB**
Fixes KDAB#22
Depends on KDAB#21

### CODE REVIEW

File: src/gui/findinfilespanel.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+#include "guisettings.h"
+#include <QFile>
+#include <QHBoxLayout>
+#include <QHeaderView>
+#include <QTreeWidgetItem>
+#include <QVBoxLayout>
+
+namespace Gui {
+
+enum FindInFilesRoles { LineRole = Qt::UserRole + 1, ColumnRole };
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:23]: No enum name needed.  `enum { ... };`

File: src/gui/findinfilespanel.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    mainLayout->setContentsMargins(0, 0, 0, 0);
+    mainLayout->setSpacing(0);
+    setupToolBar();
+    mainLayout->addWidget(m_toolBar);
+
+    m_resultsDisplay->setHeaderHidden(true);
+    m_resultsDisplay->header()->setSectionResizeMode(0, QHeaderView::Stretch);
+    mainLayout->addWidget(m_resultsDisplay);
+
+    connect(m_resultsDisplay, &QTreeWidget::itemDoubleClicked, this, [this](QTreeWidgetItem *item, int) {
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:43]: not needed, itemActivated() is emitted on double-click too

File: src/gui/findinfilespanel.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    connect(m_searchInput, &QLineEdit::returnPressed, this, [this]() {
+        findInFiles(m_searchInput->text());
+    });
+}
+
+void FindInFilesPanel::findInFiles(const QString &searchText)
+{
+    if (searchText.isEmpty()) {
+        qWarning() << "The text to search for is empty";
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:82]: That's not really worth a warning. It's not a coding error. It's a normal condition.

Alternatively, you could disable the button when the text is empty. That would probably be better.

File: src/gui/findinfilespanel.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    m_searchButton = new QToolButton(m_toolBar);
+    m_searchButton->setText(tr("Find"));
+    layout->addWidget(m_searchButton);
+
+    connect(m_searchButton, &QToolButton::clicked, this, [this]() {
+        findInFiles(m_searchInput->text());
+    });
+
+    connect(m_searchInput, &QLineEdit::returnPressed, this, [this]() {
+        findInFiles(m_searchInput->text());
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:75]: You could simplify this by removing the argument, and calling `m_searchInput()->text()` inside the method.

Then these two connects don't need a lambda anymore, you can just connect to `&FindInFilesPanel::findInFiles` as the 4th argument.

File: src/gui/findinfilespanel.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+class FindInFilesPanel : public QWidget
+{
+    Q_OBJECT
+
+public:
+    explicit FindInFilesPanel(QWidget *parent = nullptr);
+
+    QWidget *toolBar() const;
+    void displayResults(const QVariantList &results) const;
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.h:28]: Why public?

(related to the dead code further down, can be made private after removing the dead code)

File: src/gui/findinfilespanel.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+
+SPDX-License-Identifier: GPL-3.0-only
+
+Contact KDAB at <info@kdab.com> for commercial licensing options.
+*/
+
+#pragma once
+
+#include <QLineEdit>
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.h:13]: Use forward declarations instead, for anything that's only used as a pointer in this file, like QLineEdit, QToolButton, QTreeWidget

File: src/gui/mainwindow.cpp [hunk 7]
```diff
@@ -798,4 +801,9 @@ void MainWindow::changeCurrentDocument()
     updateActions();
 }
 
+void MainWindow::displayFindResults(const QVariantList &results) const
```
Comment (dfaure-kdab) [inline src/gui/mainwindow.cpp:804]: Who calls this method?

I think it's dead code.

File: src/gui/findinfilespanel.cpp [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        if (it == fileItems.end()) {
+            auto fileItem = new QTreeWidgetItem(m_resultsDisplay);
+            fileItem->setText(0, filePath);
+            it = fileItems.insert(filePath, fileItem);
+        }
+
+        auto lineItem = new QTreeWidgetItem(it.value());
+        lineItem->setText(0, QString("%1        %2").arg(line).arg(text));
+        lineItem->setData(0, LineRole, line);
+        lineItem->setData(1, ColumnRole, column);
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:133]: Set the data in column 0, not 1 (which doesn't exist)

File: src/gui/findinfilespanel.cpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+void FindInFilesPanel::openFileAtItem(QTreeWidgetItem *item)
+{
+    if (item->childCount() > 0)
+        return;
+
+    QTreeWidgetItem *parentItem = item->parent();
+    const QString filePath = parentItem->text(0);
+    const int line = item->data(0, LineRole).toInt();
+    const int column = item->data(1, ColumnRole).toInt();
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:145]: column 0 here too, obviously

File: src/gui/findinfilespanel.cpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+SPDX-License-Identifier: GPL-3.0-only
+
+Contact KDAB at <info@kdab.com> for commercial licensing options.
+*/
+
+#include "findinfilespanel.h"
+#include "core/project.h"
+#include "core/textdocument.h"
+#include "guisettings.h"
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.cpp:14]: unused?

File: src/gui/mainwindow.h [hunk 11]
```diff
@@ -123,6 +125,7 @@ class MainWindow : public QMainWindow
     ShortcutManager *const m_shortcutManager = nullptr;
     ToolBar *const m_toolBar = nullptr;
     QPointer<TreeSitterInspector> m_treeSitterInspector;
+    FindInFilesPanel *const m_findInFilesPanel = nullptr;
```
Comment (dfaure-kdab) [inline src/gui/mainwindow.h:128]: Doesn't really need to be a member variable. See logPanel for an example.

**Review by dfaure-kdab — CHANGES_REQUESTED** [review]

File: src/gui/mainwindow.h [hunk 12]
```diff
@@ -110,6 +111,7 @@ class MainWindow : public QMainWindow
     void changeCurrentDocument();
     QDockWidget *createDock(QWidget *widget, Qt::DockWidgetArea area, QWidget *toolbar = nullptr);
     void reloadDocuments();
+    void displayFindResults(const QVariantList &results) const;
```
Comment (dfaure-kdab) [inline src/gui/mainwindow.h:114]: You forgot to remove this.

File: src/gui/mainwindow.h [hunk 13]
```diff
@@ -32,6 +32,7 @@ class DocumentPalette;
 class ShortcutManager;
 class TreeSitterInspector;
 class ToolBar;
+class FindInFilesPanel;
```
Comment (dfaure-kdab) [inline src/gui/mainwindow.h:35]: You forgot to remove this.

File: src/gui/findinfilespanel.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+private:
+    void displayResults(const QVariantList &results) const;
+    void findInFiles();
+    void openFileAtItem(QTreeWidgetItem *item);
+    void setupToolBar();
+
+    QWidget *const m_toolBar;
+    QTreeWidget *m_resultsDisplay;
+    QLineEdit *m_searchInput;
+    QToolButton *m_searchButton;
```
Comment (dfaure-kdab) [inline src/gui/findinfilespanel.h:40]: Doesn't need to be a member variable.

**Review by dfaure-kdab — CHANGES_REQUESTED** [review]

**Review by dfaure-kdab — APPROVED** [review]

**Review by narnaud — APPROVED** [review]

**Pull request merged**